js/src/html/beautifier.js

```diff
@@ -229,6 +239,17 @@ function Beautifier(html_source, options, js_beautify, css_beautify) {
             }
         };
 
+        this.get_last_tag = function() {
+            var last_tag;
+            for (var i = this.output.length - 1; i >= 0; i --) {
```

Comment:
Walking back one character at a time is going to have really bad performance characteristics. 
At minimum, we should store the `last_tag` info when we find a tag and then replace it each time a new tag is found.

Reply:
This is walking back one *line* at a time, and, while I wrote it defensively, I cannot think of any situation where it would loop more than twice.